Sandia-OpenSHMEM/SOS
# Add support for shared memory atomics

This PR adds a proper shared memory transport layer (dubbed `shr_transport`).

 * Refactor all CMA and XPMEM code into a common shr transport layer
 * Add support for atomics in the shr transport layer
 * Refactor atomic datatypes to have unique values. This is needed to build switch statements on the type in the shared transport. This refactor touched a surprising amount of code, including the ops used for "software" reductions. It introduces an additional translation in the transport layer to convert the datatype to a transport-layer datatype.
 * Clean up OFI atomics checks (mostly whitespace edits)
 * Update wait/test to use atomic load with shr transport
 * Update lock implementation to use AMOs rather than directly reading the lock variable. Also found a bug where there was a missing `NEXT` around the lock value.
 * Add contexts and teams support to transport none (used now for strictly shared memory builds of SOS).


A few future work items:

 * Relax SEQ_CST ordering to ACQUIRE/RELEASE/ACQ_REL (See #923)
 * Figure out what to do about AMO reductions (currently disabled)
 * Add autoconf check for intrinsic atomics
 * Add error checks to autoconf, e.g. should --enable-shr-atomics without --with-xpmem be an error? I think this combination of flags will compile without issue, it just may not do what users expect.
